### DISCUSSION TITLE
How startup works in Nelua execution process

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#85] [Q&A] [answered]
I have noticed the following behavior: when I delete `nelua_cache` to force Nelua to regenerate the native code, its startup speed delays quite a lot.

On recompiling other sources, the startup speed drops but usually remains around 30 to 40 milliseconds approximately, give or take depending on the situation; the screenshot below is an exception though due to the nature of the code.

My question is, what does it cause the delay?

Is it compiler's cold startup on first execution?

![image](https://user-images.githubusercontent.com/1777466/127774243-eb42386f-b9b6-4745-be11-d900a094079b.png)

**Comment by edubart** [comment, marked as answer]
The compiler startup just load lua files, and check some disk files. What you are experience is probably something related to your OS, when you delete the folder probably the OS is invalidating files cache and when you startup again the OS need to recache the files. You could confirm this executing the following:
```bash
sudo sh -c "echo 3 > /proc/sys/vm/drop_caches"
```

For instance once the OS cache Nelua files, the startup takes about 20ms here, but when I drop the caches it takes more than 100ms:

```sh
➜  nelua-lang git:(master ✓) nelua -tb --no-cache examples/helloworld.nelua 
startup      22.6 ms
parse        1.5 ms
preprocess   0.0 ms
analyze      3.1 ms
generate     0.1 ms
compile      34.1 ms
total build  61.4 ms
➜  nelua-lang git:(master ✓) sync; sudo sh -c "echo 3 > /proc/sys/vm/drop_caches"
➜  nelua-lang git:(master ✓) nelua -tb --no-cache examples/helloworld.nelua 
startup      132.1 ms
parse        5.6 ms
preprocess   0.1 ms
analyze      12.0 ms
generate     0.5 ms
compile      134.9 ms
total build  285.1 ms
➜  nelua-lang git:(master ✓) nelua -tb --no-cache examples/helloworld.nelua 
startup      22.2 ms
parse        1.5 ms
preprocess   0.0 ms
analyze      3.3 ms
generate     0.1 ms
compile      34.7 ms
total build  61.8 ms
```

Notice that after dropping the OS cache Nelua takes about 4x time to compile. When the OS decides to drop caches I can't say, this vary between setups, environment and activity, this is something specific to your OS and out of the scope here. You could use high speed hard disk or even put Nelua and all its files in RAM memory to speedup.

**Reply by stefanos82** [reply]
Probably that relates somehow, even though I believe it's something else.

I have tried your example above and got these dizzying results:

![image](https://user-images.githubusercontent.com/1777466/127783998-4627055e-7e4c-4861-893c-8d8568e4b6d1.png)

**Reply by stefanos82** [reply]
> The compiler startup just load lua files, and check some disk files.

If I may ask, what files are these that get loaded anew every single executing time?

Can they get cached if nothing changes upon running Nelua on a new call, so we could reduce the unnecessary execution time upon re-execution of Nelua?

**Reply by edubart** [reply]
You can trace all files read with `strace` tool on Linux:

```
➜  nelua-lang git:(master) strace -e trace=open,openat nelua examples/helloworld.nelua
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/libreadline.so.8", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/libdl.so.2", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/libncursesw.so.6", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/dev/tty", O_RDWR|O_NONBLOCK) = 3
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/gconv/gconv-modules.cache", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/gconv/gconv-modules", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/home/bart/projects/dotfiles/bin/nelua", O_RDONLY) = 3
...
openat(AT_FDCWD, "/home/bart/projects/nelua/nelua-lang/nelua/analyzercontext.lua", O_RDONLY) = 3
openat(AT_FDCWD, "/home/bart/projects/nelua/nelua-lang/nelua/analyzercontext.lua", O_RDONLY) = 3
openat(AT_FDCWD, "/home/bart/projects/nelua/nelua-lang/nelua/scope.lua", O_RDONLY) = 3
openat(AT_FDCWD, "/home/bart/projects/nelua/nelua-lang/nelua/scope.lua", O_RDONLY) = 3
openat(AT_FDCWD, "/home/bart/.cache/nelua/examples/helloworld.c", O_RDONLY) = 3
hello world
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=117774, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
+++ exited with 0 +++
```

I've trimmed the output above, as you can see lots of file are read on startup. Usually the OS will cache these files on first file read in unused RAM memory, this is why second runs usually is faster while dropping OS cache makes everything slow again.

I've created forkmon to skip redoing that startup, so you could use `forkmon-nelua` to speedup the startup.

**Reply by stefanos82** [reply]
The interesting part of the output is the dual call for each Lua file; I really wonder why...

**Reply by edubart** [reply]
> The interesting part of the output is the dual call for each Lua file; I really wonder why...

Probably the first open is just to check if the file exists, and the second the actual read.

**Reply by stefanos82** [reply]
I see...by the way, since the change in https://github.com/edubart/nelua-lang/commit/1227c73dcb2e8778849f7c24270da6e0063aa144, the whole build and execute procedure improved exponentially.

Now I get times like 25ms and such!

**Reply by edubart** [reply]
> I see...by the way, since the change in [1227c73](https://github.com/edubart/nelua-lang/commit/1227c73dcb2e8778849f7c24270da6e0063aa144), the whole build and execute procedure improved exponentially.
> 
> Now I get times like 25ms and such!

That commit just changed a file system directory, so looks like to really be something related with OS/disk/file caching.

**Reply by stefanos82** [reply]
The thing is, if you deleted `nelua_cache` and re-executed the same code, the startup time never dropped this low before the default cache directory.

It's really incredible to be honest with you!
